AiiDA scheduler plugin thinks job is finished while it is still running

From time to time, AiiDA will retrieve the files in a calculation folder before the calculation has finished.
When checking the calculation folder using `gotocomputer` at a later point in time, one finds that the calculation actually finished fine.

From a user perspective, this issue is very similar to https://github.com/aiidateam/aiida-core/issues/3404, but we've now established (see below) that there are two different failure modes.
In the case discussed here, AiiDA gets an incorrect job list (which no longer contains the job in question) and therefore starts retrieving it.

===========

@sphuber We ran into this issue on helvetios/fidis.

Now, following up on your suggestion to investigate the failure mode, I have looked at the historical job records of both clusters during that time using `sacct`. In particular, one can query for the job ids and work directories of past jobs as follows:

```bash
# get work dirs of jobs since 2020-07-01 
sacct -S 2020-07-01 -u ongari --format "jobid,jobname%20,workdir%70" > sacct-list
cat sacct-list | grep scratch | grep aiida_run | awk '{print $3}' > directories
cat directories | wc -l  # prints 4705
cat directories | sort -u | wc -l  # prints 4705
```

It turned out that the list of work directories is unique.
If the reason for the failure was that AiiDA submitted the same job twice (causing one or both of them to fail), then we should see the corresponding work directories appear twice in the list.
I.e. in this particular case, I believe the issue occurs at the point where AiiDA queries the scheduler for the job status and gets an incomplete/incorrect answer.

As mentioned during our discussion, one solution of this issue could be to add a fail-safe mechanism when AiiDA gets the information that a job has completed. This could be any of
 * running `squeue` again, just to be sure
 * using an alternative command, that queries the status of that job specifically, such as
   * `scontrol show job <id>` 
   * `sacct -j <id>` 

and only accepting the fact that the job is no longer running, if the output agrees.

Mentioning @danieleongari for info

_Originally posted by @ltalirz in https://github.com/aiidateam/aiida-core/issues/3404#issuecomment-672951815_

From our discussion, possible explanations include:
 1) for whatever reason, the output of squeue is too long for the ssh buffer (~4 MB according to @giovannipizzi )
This is possible but unlikely since squeue tries to query the jobs per user
https://github.com/aiidateam/aiida-core/blob/aa9a2cb519f96fef24746a7ffb8e5701107f2503/aiida/schedulers/plugins/slurm.py#L206-L207
 2) perhaps more likely: the output from squeue is incomplete and still parsed by aiida

Re 2. 
It turns out that AiiDA gets the return value of the joblist command and forwards it to the `_parse_joblist_output` function of the scheduler plugin, but the function for slurm ignores a nonzero exit status if stderr is empty
https://github.com/aiidateam/aiida-core/blob/aa9a2cb519f96fef24746a7ffb8e5701107f2503/aiida/schedulers/plugins/slurm.py#L485-L499

This seems a dangerous practice to me and would probably be quite straightforward to improve.

P.S. Unfortunately, we don't have the complete daemon logs from the time of the failed calculation anymore to check exactly what was printed there.
Searching through the recent log history, however, it does seem that this part of the code *does* catch socket timeouts on `squeue` from time to time, e.g. : 
```
08/21/2020 09:12:45 AM <26803> aiida.scheduler.slurm: [WARNING] Warning in _parse_joblist_output, non-empty stderr='slurm_load_jobs error: Socket timed out on send/recv operation'
08/21/2020 09:12:45 AM <26803> aiida.engine.transports: [ERROR] Exception whilst using transport:
Traceback (most recent call last):
  File "/home/daniele/aiida1/aiida_core/aiida/engine/transports.py", line 103, in request_transport
    yield transport_request.future
  File "/home/daniele/aiida1/aiida_core/aiida/engine/processes/calcjobs/manager.py", line 106, in _get_jobs_from_scheduler
    scheduler_response = scheduler.get_jobs(**kwargs)
  File "/home/daniele/aiida1/aiida_core/aiida/schedulers/scheduler.py", line 340, in get_jobs
    joblist = self._parse_joblist_output(retval, stdout, stderr)
  File "/home/daniele/aiida1/aiida_core/aiida/schedulers/plugins/slurm.py", line 499, in _parse_joblist_output
    raise SchedulerError('Error during squeue parsing (_parse_joblist_output function)')
aiida.schedulers.scheduler.SchedulerError: Error during squeue parsing (_parse_joblist_output function)
```
In these cases, stderr was non-empty and the scheduler error was raised as designed - but perhaps under high load / special circumstances, stderr can remain empty?

--- a/aiida/schedulers/plugins/slurm.py
+++ b/aiida/schedulers/plugins/slurm.py
@@ -228,7 +228,7 @@ class SlurmScheduler(Scheduler):
             # Verified on slurm versions 17.11.2, 19.05.3-2 and 20.02.2.
             # See also https://github.com/aiidateam/aiida-core/issues/4326
             if len(joblist) == 1:
-                joblist += joblist[0]
+                joblist += [joblist[0]]
 
             command.append('--jobs={}'.format(','.join(joblist)))
 

--- a/aiida/schedulers/plugins/slurm.py
+++ b/aiida/schedulers/plugins/slurm.py
@@ -214,6 +214,19 @@ class SlurmScheduler(Scheduler):
                 if not isinstance(jobs, (tuple, list)):
                     raise TypeError("If provided, the 'jobs' variable must be a string or a list of strings")
                 joblist = jobs
+
+            # Trick: If we are asking for a single job, we append the same job once more.
+            # We do this because squeue returns a non-zero exit code if provided an invalid jobid
+            # (stderr: "slurm_load_jobs error: Invalid job id specified"), which can happen if time has passed since
+            # the job finished and squeue no longer has knowledge of it.
+            # However, this *only* happens when providing a *single* jobid to squeue.
+            # When providing two or more jobids via `squeue --jobs=123,234`, it always returns exit code zero, and
+            # duplicating job ids has no other effect on the output.
+            # Verified on slurm versions 17.11.2, 19.05.3-2 and 20.02.2.
+            # See also https://github.com/aiidateam/aiida-core/issues/4326
+            if len(joblist) == 1:
+                joblist += joblist[0]
+
             command.append('--jobs={}'.format(','.join(joblist)))
 
         comm = ' '.join(command)
@@ -482,21 +495,18 @@ class SlurmScheduler(Scheduler):
         # pylint: disable=too-many-branches,too-many-statements
         num_fields = len(self.fields)
 
-        # I don't raise because if I pass a list of jobs,
-        # I get a non-zero status
-        # if one of the job is not in the list anymore
-        # retval should be zero
-        # if retval != 0:
-        # self.logger.warning("Error in _parse_joblist_output: retval={}; "
-        #    "stdout={}; stderr={}".format(retval, stdout, stderr))
-
-        # issue a warning if there is any stderr output and
-        # there is no line containing "Invalid job id specified", that happens
-        # when I ask for specific calculations, and they are all finished
-        if stderr.strip() and 'Invalid job id specified' not in stderr:
-            self.logger.warning("Warning in _parse_joblist_output, non-empty stderr='{}'".format(stderr.strip()))
-            if retval != 0:
-                raise SchedulerError('Error during squeue parsing (_parse_joblist_output function)')
+        if retval != 0:
+            raise SchedulerError(
+                """squeue returned exit code {} (_parse_joblist_output function)
+stdout='{}'
+stderr='{}'""".format(retval, stdout.strip(), stderr.strip())
+            )
+        if stderr.strip():
+            self.logger.warning(
+                "squeue returned exit code 0 (_parse_joblist_output function) but non-empty stderr='{}'".format(
+                    stderr.strip()
+                )
+            )
 
         # will contain raw data parsed from output: only lines with the
         # separator, and already split in fields

--- a/aiida/schedulers/plugins/slurm.py
+++ b/aiida/schedulers/plugins/slurm.py
@@ -215,13 +215,16 @@ class SlurmScheduler(Scheduler):
                     raise TypeError("If provided, the 'jobs' variable must be a string or a list of strings")
                 joblist = jobs
 
-            # Trick: If we are asking for a single job, we append the same job once more.
-            # We do this because squeue returns a non-zero exit code if provided an invalid jobid
-            # (stderr: "slurm_load_jobs error: Invalid job id specified"), which can happen if time has passed since
-            # the job finished and squeue no longer has knowledge of it.
-            # However, this *only* happens when providing a *single* jobid to squeue.
-            # When providing two or more jobids via `squeue --jobs=123,234`, it always returns exit code zero, and
-            # duplicating job ids has no other effect on the output.
+            # Trick: When asking for a single job, append the same job once more.
+            # This helps provide a reliable way of knowing whether the squeue command failed (if its exit code is
+            # non-zero, _parse_joblist_output assumes that an error has occurred and raises an exception).
+            # When asking for a single job, squeue also returns a non-zero exit code if the corresponding job is
+            # no longer in the queue (stderr: "slurm_load_jobs error: Invalid job id specified"), which typically
+            # happens once in the life time of an AiiDA job,
+            # However, when providing two or more jobids via `squeue --jobs=123,234`, squeue stops caring whether
+            # the jobs are still in the queue and returns exit code zero irrespectively (allowing AiiDA to rely on the
+            # exit code for detection of real issues).
+            # Duplicating job ids has no other effect on the output.
             # Verified on slurm versions 17.11.2, 19.05.3-2 and 20.02.2.
             # See also https://github.com/aiidateam/aiida-core/issues/4326
             if len(joblist) == 1:
@@ -495,6 +498,7 @@ class SlurmScheduler(Scheduler):
         # pylint: disable=too-many-branches,too-many-statements
         num_fields = len(self.fields)
 
+        # See discussion in _get_joblist_command on how we ensure that AiiDA can expect exit code 0 here.
         if retval != 0:
             raise SchedulerError(
                 """squeue returned exit code {} (_parse_joblist_output function)

--- a/tests/schedulers/test_slurm.py
+++ b/tests/schedulers/test_slurm.py
@@ -14,6 +14,7 @@ import uuid
 import datetime
 
 from aiida.schedulers.plugins.slurm import SlurmScheduler, JobState
+from aiida.schedulers import SchedulerError
 
 # pylint: disable=line-too-long
 # job_id, state_raw, annotation, executing_host, username, number_nodes, number_cpus, allocated_machines, partition, time_limit, time_used, dispatch_time, job_name, submission_time
@@ -42,7 +43,7 @@ class TestParserSqueue(unittest.TestCase):
 
     def test_parse_common_joblist_output(self):
         """
-        Test whether _parse_joblist can parse the qstat -f output
+        Test whether _parse_joblist_output can parse the squeue output
         """
         scheduler = SlurmScheduler()
 
@@ -98,6 +99,19 @@ class TestParserSqueue(unittest.TestCase):
         #
         #                self.assertTrue( j.num_machines==num_machines )
         #                self.assertTrue( j.num_mpiprocs==num_mpiprocs )
+    def test_parse_failed_squeue_output(self):
+        """
+        Test that _parse_joblist_output reacts as expected to failures.
+        """
+        scheduler = SlurmScheduler()
+
+        # non-zero return value should raise
+        with self.assertRaises(SchedulerError):
+            _ = scheduler._parse_joblist_output(1, TEXT_SQUEUE_TO_TEST, '')  # pylint: disable=protected-access
+
+        # non-empty stderr should be logged
+        with self.assertLogs(scheduler.logger, 'WARNING'):
+            _ = scheduler._parse_joblist_output(0, TEXT_SQUEUE_TO_TEST, 'error message')  # pylint: disable=protected-access
 
 
 class TestTimes(unittest.TestCase):
@@ -332,3 +346,24 @@ class TestSubmitScript(unittest.TestCase):
             job_tmpl.job_resource = scheduler.create_job_resource(
                 num_machines=1, num_mpiprocs_per_machine=1, num_cores_per_machine=24, num_cores_per_mpiproc=23
             )
+
+
+class TestJoblistCommand(unittest.TestCase):
+    """
+    Tests of the issued squeue command.
+    """
+
+    def test_joblist_single(self):
+        """Test that asking for a single job results in duplication of the list."""
+        scheduler = SlurmScheduler()
+
+        command = scheduler._get_joblist_command(jobs=['123'])  # pylint: disable=protected-access
+        self.assertIn('123,123', command)
+
+    def test_joblist_multi(self):
+        """Test that asking for multiple jobs does not result in duplications."""
+        scheduler = SlurmScheduler()
+
+        command = scheduler._get_joblist_command(jobs=['123', '456'])  # pylint: disable=protected-access
+        self.assertIn('123,456', command)
+        self.assertNotIn('456,456', command)

